irohad/pending_txs_storage/impl/pending_txs_storage_impl.hpp

```diff
@@ -0,0 +1,77 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef IROHA_PENDING_TXS_STORAGE_IMPL_HPP
+#define IROHA_PENDING_TXS_STORAGE_IMPL_HPP
+
+#include <set>
+#include <shared_mutex>
+#include <unordered_map>
+#include <unordered_set>
+
+#include <rxcpp/rx.hpp>
+#include "pending_txs_storage/pending_txs_storage.hpp"
+#include "interfaces/iroha_internal/transaction_batch.hpp"
+
+namespace iroha {
+
+  class MstState;
+
+  class PendingTransactionStorageImpl : public PendingTransactionStorage {
+   public:
+    using AccountIdType = shared_model::interface::types::AccountIdType;
+    using HashType = shared_model::interface::types::HashType;
+    using SharedTxsCollectionType =
+        shared_model::interface::types::SharedTxsCollectionType;
+    using TransactionBatch = shared_model::interface::TransactionBatch;
+    using SharedState = std::shared_ptr<MstState>;
+    using SharedBatch = std::shared_ptr<TransactionBatch>;
+    using StateObservable = rxcpp::observable<SharedState>;
+    using BatchObservable = rxcpp::observable<SharedBatch>;
+
+    PendingTransactionStorageImpl(StateObservable updatedBatches,
+                                  BatchObservable preparedBatch,
+                                  BatchObservable expiredBatch);
+
+    ~PendingTransactionStorageImpl() override;
+
+    SharedTxsCollectionType getPendingTransactions(
+        const AccountIdType &accountId) const override;
+
+   private:
+    void updatedBatchesHandler(const SharedState &updatedBatches);
+    void preparedBatchHandler(const SharedBatch &preparedBatch);
+    void expiredBatchHandler(const SharedBatch &expiredBatch);
+
+    void removeBatch(const SharedBatch &batch);
+
+    std::set<AccountIdType> batchCreators(const TransactionBatch &batch) const;
+
+    /**
+     * Subscriptions on MST events
+     */
+    rxcpp::composite_subscription updated_batches_subscription_;
+    rxcpp::composite_subscription prepared_batch_subscription_;
+    rxcpp::composite_subscription expired_batch_subscription_;
+
+    /**
+     * Mutex for single-write multiple-read storage access
+     */
+    mutable std::shared_timed_mutex mutex_;
+
+    struct {
```

Comment:
This struct and purpose unobvious for me. Pls, add documentation for class(struct) and fields with explanation HOW it works.